src/hazardous/kdf/scrypt.rs

```diff
@@ -0,0 +1,445 @@
+    x15 = x15.wrapping_add(w15);
+
+    out[0] = x0; tmp[0] = x0;
+    out[1] = x1; tmp[1] = x1;
+    out[2] = x2; tmp[2] = x2;
+    out[3] = x3; tmp[3] = x3;
+    out[4] = x4; tmp[4] = x4;
+    out[5] = x5; tmp[5] = x5;
+    out[6] = x6; tmp[6] = x6;
+    out[7] = x7; tmp[7] = x7;
+    out[8] = x8; tmp[8] = x8;
+    out[9] = x9; tmp[9] = x9;
+    out[10] = x10; tmp[10] = x10;
+    out[11] = x11; tmp[11] = x11;
+    out[12] = x12; tmp[12] = x12;
+    out[13] = x13; tmp[13] = x13;
+    out[14] = x14; tmp[14] = x14;
+    out[15] = x15; tmp[15] = x15;
+}
+
+#[rustfmt::skip]
+fn block_mix(tmp: &mut [u32], inn: &[u32], out: &mut [u32], r: usize) {
+    block_copy(tmp, &inn[(2*r-1)*16..], 16);
+    for i in (0..2*r).step_by(2) {
+        salsa_xor(tmp, &inn[i*16..], &mut out[i*8..]);
+        salsa_xor(tmp, &inn[i*16+16..], &mut out[i*8+r*16..]);
+    }
+}
+
+fn integer(b: &[u32], r: usize) -> u64 {
+    let j = (2 * r - 1) * 16;
+    u64::from(b[j]) | u64::from(b[j + 1]) << 32
+}
+
+#[allow(non_snake_case)]
+fn smix(b: &mut [u8], r: usize, N: usize, v: &mut [u32], x: &mut [u32], y: &mut [u32]) {
+    let mut tmp = [0u32; 16];
+    let R = 32 * r;
+
+    let mut j = 0;
+    for i in 0..R {
+        x[i] = u32::from_le_bytes(b[j..j + 4].try_into().unwrap());
+        j += 4;
+    }
+
+    for i in (0..N as usize).step_by(2) {
+        block_copy(&mut v[i * R..], x, R);
+        block_mix(&mut tmp, x, y, r);
+
+        block_copy(&mut v[(i + 1) * R..], y, R);
+        block_mix(&mut tmp, y, x, r);
+    }
+
+    for _ in (0..N).step_by(2) {
+        let j = (integer(x, r) & u64::from((N - 1) as u64)) as usize;
+        block_xor(x, &v[j * R..], R);
+        block_mix(&mut tmp, x, y, r);
+
+        let j = (integer(y, r) & u64::from((N - 1) as u64)) as usize;
+        block_xor(y, &v[j * R..], R);
+        block_mix(&mut tmp, y, x, r);
+    }
+
+    let mut j = 0;
+    for v in &x[..R] {
+        b[j..j + 4].copy_from_slice(&v.to_le_bytes());
+        j += 4;
+    }
+}
+
+#[must_use = "SECURITY WARNING: Ignoring a Result can have real security implications."]
+/// scrypt key derivation function as specified in RFC 7914
+pub fn derive_key(
+    password: &[u8],
+    salt: &[u8],
+    n: u32,
+    r: u32,
+    p: u32,
+    dst_out: &mut [u8],
+) -> Result<(), UnknownCryptoError> {
+    if !(n > 1)
+        || !(n & (n - 1) == 0)
+        || !(r * p < 1 << 30)
+        || !(r <= (i32::MAX as u32) / 128 / p)
+        || !(r <= (i32::MAX as u32) / 256)
+        || !(n <= (i32::MAX as u32) / 128 / r)
+        || !(usize::BITS >= 32)
+    {
+        return Err(UnknownCryptoError);
+    }
+
+    let n: usize = n as usize;
+    let r: usize = r as usize;
+    let p: usize = p as usize;
+
+    let vlen: usize = 32 * n * r;
+    let mut x = vec![0u32; 32 * r];
+    let mut y = vec![0u32; 32 * r];
+    let mut v = vec![0u32; vlen];
+    let pass = pbkdf2::Password::from_slice(password)?;
+    let blen: usize = p * 128 * r;
+    let mut b = vec![0u8; blen];
```

Comment:
Should this `Vec<u8>` be `Zeroizing::new(...)`? It's the weakest hash possible of a sensitive user-supplied password, so if the performance penalty isn't outrageous, maybe it makes sense to have it zeroed after it goes out of scope?